File: src/main/java/core/procedures/symbols/StringToSymbol.java
package core.procedures.symbols;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMSymbol;

public class StringToSymbol extends AFn {

  public StringToSymbol() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{CharSequence.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "string->symbol";
  }

  @Override
  public SCMSymbol apply1(Object arg) {
    return SCMSymbol.intern(arg.toString());
  }
}

File: src/main/java/core/procedures/symbols/StringToSymbol.kt
package core.procedures.symbols

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Symbol

open class StringToSymbol : AFn<CharSequence?, Symbol>(name = "string->symbol", isPure = true, arity = Exactly(1),
                                                       mandatoryArgsTypes = arrayOf(CharSequence::class.java)) {

    override operator fun invoke(arg: CharSequence?) = Symbol.intern(arg!!.toString())
}
